### ISSUE TITLE
Regression: REST proxy no longer accept the offical json encoding of Avro value.

### ISSUE THREAD

**Issue by perj** [aiven/karapace#319]
# What happened?

We activated karapace on one of our clusters today.
One service producing using the REST proxy started having 42205 errors, indicating the values doesn't match the schema.
It didn't have these errors before we switched to Karapace.

Weirdly I couldn't reproduce the problem on another cluster we switched earlier. So I updated that cluster, and then the appeared there as well. That indicates this is a regression.

# What did you expect to happen?

Given this schema upload:

```
curl https://karapace-schema-registry/subjects/test/versions -H 'Content-Type: application/json' -d '{"schema":"[\"null\", \"boolean\"]"}'
{"id":205}
```

I expected this to work:

```
curl -v https://karapace-rest-proxy/topics/test -H 'Content-Type: application/vnd.kafka.avro.v2+json' -d '{"value_schema_id":205,"records":[{"value":{"boolean":true}}]}'
```

Before my cluster upgrade to the latest karapace, it did work:
```
{"key_schema_id":null,"offsets":[{"offset":0,"partition":0}],"value_schema_id":205}
```

But after the upgrade it does not:
```
{"error_code":42205,"message":"Object does not fit to stored schema"}
```

Also, I would expect this to NOT work, although I'm ok if both versions work I suppose.
```
curl -v https://karapace-rest-proxy/topics/test -H 'Content-Type: application/vnd.kafka.avro.v2+json' -d '{"value_schema_id":205,"records":[{"value":true}]}'
```

It didn't work before the upgrade:
```
{"error_code":42205,"message":"true is no instance of ['null', 'boolean']"}
```

but after the upgrade it does work, unexpectedly:
```
{"key_schema_id":null,"offsets":[{"offset":1,"partition":0}],"value_schema_id":205}
```

# What else do we need to know?

Seems likely to be due to this commit: https://github.com/aiven/karapace/commit/2321042bb080428c8b5e7cf20eb65e371e3913d5

**Comment by hackaugusto** [comment]
Hi @perj , thank you for the report.

We do have tests for this scenarion [here](https://github.com/aiven/karapace/blob/de400baf1ebcc0832afe932d3fe9026276dc73a3/tests/integration/test_rest.py#L37). I'm going to double check if they are covering the same thing and working as expected.

**Comment by hackaugusto** [comment]
@perj could you please confirm the old version that worked with this content type? The issue seems to be the `v2`.

**Comment by perj** [comment]
Hi. Thanks for the quick response.

I'm not sure how to check the versions I'm afraid... I'm only using Karapace through the Aiven hosted Kafka.
I've checked the logs but it doesn't state the Karapace version as far as I can tell.

Like I mentioned in my edit, it seems likely to be introduced by https://github.com/aiven/karapace/commit/2321042bb080428c8b5e7cf20eb65e371e3913d5

**Comment by perj** [comment]
The previous time an maintenance upgrade was done on the cluster that worked before upgrading yesterday was on Dec 13 if that helps.

**Comment by perj** [comment]
The problem only applies to union types. I might be missing it, but I don't see any tests for union types in your link.

**Comment by hackaugusto** [comment]
TL;DR This is a regression, union values are not handled properly.

---

> The problem only applies to union types. I might be missing it, but I don't see any tests for union types in your link.

Ah, okay. I misunderstood the problem in the first read, this is indeed a regression. Karapace has a known bug, it doesn't properly handle Avro union types. Unions have a special rule for JSON encoding [Avro spec](https://avro.apache.org/docs/current/spec.html#json_encoding):

> Except for unions, the JSON encoding is the same as is used to encode field default values. [...]

[Karapace](https://github.com/aiven/karapace/blob/41c7c94fecd6a4a808f9e64194eacd8e11344bbc/karapace/avro_compatibility.py#L36-L53) and the version of the avro library used are not handling that.

The bug was fixed by https://github.com/aiven/karapace/pull/285, this was released together with https://github.com/aiven/karapace/pull/294 which introduced validation for schema's default value. The stricter (but correct) validation made Karapace reject invalid schemas registered via Schema Registry, which broke producers and consumers :( 

So the changes for the _release_ were reverted, because of that REST Proxy can't perform writes for schemas with union types. We will be introducing the fixes back once we handle the corner cases.
